### PR TITLE
[dev] Add Megatron-FSDP v2 prototype

### PR DESCRIPTION
**Pull request by shjwudp**
- [x] I, the PR author, have personally reviewed every line of this PR.

> Companion PR on main branch: https://github.com/NVIDIA/Megatron-LM/pull/5990

## What does this PR do?

This PR introduces the **Megatron-FSDP v2 prototype**, a DTensor-backed fully
sharded data-parallel implementation designed for Megatron Core workloads.
It is experimental and is not yet intended to replace the existing
Megatron-FSDP implementation as the production default.

The prototype provides a PyTorch FSDP2-style `fully_shard()` API while adding
execution paths optimized for Megatron models: communication overlap, Megatron
distributed-optimizer integration, hybrid sharded data parallelism,
low-precision training, distributed checkpointing, and module-level CUDA
Graphs.

## Key features

- DTensor-backed parameters, gradients, and optimizer states.
- A composable `fully_shard()` API with DDP and ZeRO-equivalent strategies:
  - `no_shard`: replicated parameters, gradients, and optimizer states.
  - `optim`: sharded optimizer states.
  - `optim_grads`: sharded gradients and optimizer states.
  - `optim_grads_params`: fully sharded parameters, gradients, and optimizer
    states.
- HSDP using a two-dimensional data-parallel mesh.
- BF16, FP8, NVFP4, and MXFP8 mixed-precision policies.
- Parameter all-gather prefetching and asynchronous gradient reduction.
- Lazy distributed-gradient allocation and reusable gradient DTensor views.
- Trace-pool allocation for stable buffer addresses.
- Experimental leaf-module CUDA Graph capture through
  `fully_shard(..., enable_cuda_graph=True)`.
- Meta-device initialization and leaf-to-root materialization.
- Megatron distributed-checkpoint support, including uneven DTensor metadata.
- Integration with Megatron optimizers, Muon, MoE, expert parallelism, and
  1F1B expert-parallel communication overlap.
- Prototype examples covering PyTorch FSDP comparison, HSDP, CUDA Graphs,
  checkpointing, QwenImage, and Qwen3/MXFP8 training.

Megatron-FSDP v2 is enabled in the Megatron training stack with:

```text
--use-megatron-fsdp-v2
```

This option also enables the common Megatron-FSDP integration path.

## Optimizing the PyTorch-native `fully_shard` model

The standalone `fsdp_toy` example compares PyTorch FSDP2 with
Megatron-FSDP v2 using the same model and full-sharding policy. Each result was
measured on 4x GB200 GPUs for 100 steps after warmup.

| Backend | Average step | Samples/s | Peak allocated |
| --- | ---: | ---: | ---: |
| PyTorch FSDP2 `fully_shard`, full shard | 13.05 ms | 2,453 | **0.96 GB** |
| Megatron-FSDP v2, full shard | 11.57 ms | 2,766 | 1.01 GB |
| Megatron-FSDP v2, `enable_cuda_graph=True` | **9.22 ms** | **3,471** | 1.32 GB |
| Megatron-FSDP v2, HSDP | 12.12 ms | 2,640 | 1.16 GB |

Compared with PyTorch FSDP2, eager Megatron-FSDP v2 improves throughput by
12.8%. Enabling the built-in module CUDA Graph path improves throughput by
41.5% and reduces average step time by 29.3% relative to PyTorch FSDP2.

The deterministic BF16 correctness test also passes for full sharding, CUDA
Graph execution, and HSDP. After the HSDP replica-refresh fix, the HSDP final
loss differs from eager full sharding by `8e-9`.

## QwenImage: module CUDA Graph performance

The production-shaped example uses the pretrained 60-block
`QwenImageTransformer2DModel`. The CUDA Graph configuration applies
`fully_shard(..., enable_cuda_graph=True)` to each transformer block; the root
module remains eager because a parent containing captured FSDP children cannot
also be captured.

The matched benchmark used 4x GB200 GPUs, full sharding, a per-rank batch size
of 4, 512x512 images, BF16, FlashAttention 2, per-block `torch.compile`, the
real-data training path, three warmup steps, and 20 measured steps.
These results were collected at `mfsdp_refactor` commit `31334f8807d6`.

| Backend | Average step | Median step | Peak memory | Final / initial loss |
| --- | ---: | ---: | ---: | ---: |
| PyTorch FSDP1 | 516.03 ms | 491.23 ms | 75.39 GB | 0.642 |
| Megatron-FSDP v2 eager | 505.89 ms | 506.46 ms | **74.67 GB** | 0.634 |
| Megatron-FSDP v2 + `enable_cuda_graph=True` | **385.89 ms** | **385.53 ms** | 86.72 GB | 0.641 |

Eager Megatron-FSDP v2 uses 0.72 GB less peak memory than PyTorch FSDP1 and is
1.96% faster by average step. Module CUDA Graph execution improves the v2
average step by 23.72% and median step by 23.88%, at the cost of 12.05 GB of
additional pool-backed memory. It is 25.22% faster than PyTorch FSDP1 by
average step.

CUDA Graph capture completes for all 60 transformer blocks, and all measured
iterations finish without a trace-pool slot collision. All three real-data
runs pass the 20-step convergence threshold.

## Alignment with existing Megatron Core parallelism

The full-shape Qwen3.5-VL 35B-A3B comparison used two nodes with four GB200
GPUs per node. All backends used the full 40-layer model, BF16,
TP1/PP1/CP1/EP8/ETP1, sequence length 4096, MBS8/GBS384, gradient-accumulation
fusion, and forced-balanced routing. Megatron-FSDP v1 and v2 used HybridEP;
the tuned ND-parallel reference used its stable NCCL all-to-all dispatcher.

Each performance run contains 50 iterations, with the first 10 excluded from
the steady-state aggregate. The matched Megatron-FSDP v1/v2 pair was collected
at commit `d1f1f68bb549`.

| Backend | Median step | Median TFLOP/s/GPU | Samples/s | Peak device memory |
| --- | ---: | ---: | ---: | ---: |
| Megatron-FSDP v1 + HybridEP | **12,258.5 ms** | **359.6** | **31.22** | **174,822 MB** |
| Megatron-FSDP v2 + HybridEP | 12,265.1 ms | 359.4 | 30.96 | 175,480 MB |
| ND parallel + NCCL all-to-all | 12,424.7 ms | 354.8 | 30.86 | 184,620 MB |

Megatron-FSDP v2 is at steady-state parity with Megatron-FSDP v1: median
throughput differs by only 0.06%. Compared with the tuned ND-parallel
configuration, v2 is 1.3% faster and uses 9,140 MB less peak device memory.
This comparison uses each backend's tuned and supported MoE transport; it is
not a dispatcher-only ablation.

### CORD-v2 convergence and ND checkpoint handoff

The full-shape Qwen3.5-VL 35B-A3B model was first trained on CORD-v2 with ND
parallelism for 200 optimizer steps using GBS16 and gradient-accumulation
fusion. The resulting ND checkpoint was then loaded into two matched
continuation runs:

- ND parallelism for 500 additional optimizer steps.
- Megatron-FSDP v2 for 500 additional optimizer steps.

Both continuations start from the same step-200 checkpoint and reach global
step 700.

Qwen3.5-VL 35B-A3B CORD-v2 convergence after ND checkpoint handoff

The post-handoff ND and Megatron-FSDP v2 loss curves begin at the same
step-200 endpoint and remain closely aligned throughout the 500-step
continuation. Both continue decreasing smoothly, with no loss discontinuity
at checkpoint load. This validates stable real-data convergence and
ND-to-Megatron-FSDP v2 checkpoint interoperability for the full model; it is
not intended as a deterministic backend-parity test.

Full W&B report:
[Qwen3.5-VL 35B-A3B CORD-v2 convergence — ND checkpoint handoff to Megatron-FSDP v2 and ND](https://wandb.ai/nvidia/qwen35-vl-35b-a3b-GB200-benchmark/reports/Qwen3-5-VL-35B-A3B-CORD-v2-convergence-ND-checkpoint-handoff-to-MFSDP-v2-ND--VmlldzoxNzU1MjU4Mw).

## Focused validation

- Megatron-FSDP v2 focused unit suite: **93 passed, 6 skipped**.
- ND/Megatron-FSDP full-iteration CUDA Graph end-to-end suite:
  **9 passed, 6 skipped**.
- Real-data convergence checks completed for QwenImage, Qwen3-30B-A3B, and
  full-shape Qwen3.5-VL 35B-A3B.

## Status

Megatron-FSDP v2 is introduced as an experimental prototype and does not
replace the existing Megatron-FSDP implementation or change the default FSDP
backend. Unsupported FSDP2 options and current checkpoint limitations are
documented in the v2 README and design documents.

# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by shjwudp** [comment]
@BestJuly suggested a more in-depth convergence validation for Qwen3.5, including post-training validation.

**Comment by shjwudp** [comment]
@FDecaYed suggested splitting the work into smaller, focused PRs, potentially moving the MuonOptimizer changes into a separate PR.

**Comment by shjwudp** [comment]
> @FDecaYed suggested splitting the work into smaller, focused PRs, potentially moving the MuonOptimizer changes into a separate PR.

@Autumn1998, could you please take a look at this issue?

**Comment by shjwudp** [comment]
> @BestJuly suggested a more in-depth convergence validation for Qwen3.5, including post-training validation.

Post-training validation results look good ✅

The convergence curve is attached below for reference:

Qwen3.5 post-training convergence

Full W&B report: [Qwen3.5-VL-35B-A3B CORD-v2 Convergence — ND Checkpoint Handoff to MFSDP v2 ND](https://wandb.ai/nvidia/qwen35-vl-35b-a3b-GB200-benchmark/reports/Qwen3-5-VL-35B-A3B-CORD-v2-convergence-ND-checkpoint-handoff-to-MFSDP-v2-ND--VmlldzoxNzU1MjU4Mw)
